### ISSUE TITLE
Wrong contents written for particular binary file

### ISSUE THREAD

**Issue by carrotIndustries** [stlink-org/stlink#193]
When writing this https://www.dropbox.com/s/ykshejjyu77pwmy/rand.bin file at 0x8004000 using

```
st-flash  write /tmp/rand.bin 0x8004000
```

and reading it back using

```
st-flash read /tmp/flash.bin 0x8004000 0x7d0
```

the read data https://www.dropbox.com/s/9w2vjmo1ap0i97e/flash.bin differs from the written data around offset 0x650.  Writing the binary file using the built-in dfu bootloader works just fine.

STM320F407 on STM32F4DISCOVERY

**Comment by xor-gate** [comment]
As this is a old issue, please test against latest tools in master, and latest stlink firmware. Reopen a new issue when the problem still exists with more details.

**Comment by Nightwalker-87** [comment]
Reopening for final verification/testing with Release v1.6.0.

**Comment by Nightwalker-87** [comment]
@carrotIndustries: I am currently reviewing old leftovers.
Is this resolved with Release v1.2.0? Please verify.

**Comment by Nightwalker-87** [comment]
This problem appeared about one month before Release v1.0.0. This has likely been fixed somewhen around that point in time, but not clear where exactly - which would be helpful though. Maybe addressed by 5dcd45d8a48aaead2a2d1d1b9c36ab14f507628b ? Any helpful ideas when looking at it?

**Comment by Nightwalker-87** [comment]
@chenguokai: You contributed to "st-flash write is ignoring 0xff at end of file" (Issue 609).
Could this be related to optimisation too?

**Comment by chenguokai** [comment]
It could be, I will check the binary file if possible.

**Comment by Nightwalker-87** [comment]
Yes, that's true - though optimisation may also be applied in other parts than the end.
That's why I came up with this, not to compare it with the 0xff-issue exactly.

**Comment by Nightwalker-87** [comment]
Obviously fixed by #377 where the mentioned code section was rewritten.
According to https://github.com/stlink-org/stlink/pull/377#issuecomment-189691156 the previous implementation was not functional.
